File: liv/src/main/java/com/benitobertoli/liv/Liv.java
package com.benitobertoli.liv;

import android.support.design.widget.TextInputLayout;

import com.benitobertoli.liv.rule.Rule;
import com.benitobertoli.liv.validator.MessageType;
import com.benitobertoli.liv.validator.TextInputLayoutValidator;
import com.benitobertoli.liv.validator.ValidationTime;
import com.benitobertoli.liv.validator.Validator;
import com.benitobertoli.liv.validator.ValidatorState;
import com.jakewharton.rxrelay2.PublishRelay;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;


public class Liv {
    private List<Validator> validators;
    private ValidatorState state = ValidatorState.NOT_VALIDATED;
    private boolean submitWhenValid = false;
    private Callback callback;
    private Action submitAction;
    private Disposable livDisposable;

    private Liv() {
        // prevent instantiation outside of the object
    }

    private Liv(Builder builder) {
        validators = builder.validators;
        callback = builder.callback;
        submitAction = builder.submitAction;
    }

    public void start() {
        ArrayList<PublishRelay<ValidatorState>> relays = new ArrayList<>(validators.size());
        for (Validator validator : validators) {
            relays.add(validator.getStateRelay());
        }

        livDisposable = Observable.combineLatest(relays,
                states -> {
                    boolean hasNotValidated = false;
                    boolean hasInvalid = false;

                    for (Object obj : states) {
                        ValidatorState state = (ValidatorState) obj;
                        if (state == ValidatorState.NOT_VALIDATED) {
                            hasNotValidated = true;
                        } else if (state == ValidatorState.VALIDATING) {
                            // one of the validator is currently validating
                            return ValidatorState.VALIDATING;
                        } else if (state == ValidatorState.INVALID) {
                            hasInvalid = true;
                        }
                    }

                    if (hasNotValidated) {
                        return ValidatorState.NOT_VALIDATED;
                    } else if (hasInvalid) {
                        return ValidatorState.INVALID;
                    }

                    return ValidatorState.VALID;
                })
                .subscribe(validatorState -> {
                    state = validatorState;
                    if (callback != null) {
                        callback.onStateChange(validatorState);
                    }

                    if (submitAction != null) {
                        if (submitWhenValid) {
                            if (state == ValidatorState.VALID) {
                                submitWhenValid = false;
                                submitAction.performAction();
                            } else if (state == ValidatorState.INVALID) {
                                submitWhenValid = false;
                            }
                        }
                    }
                }, Throwable::printStackTrace);
    }

    public void validate() {
        for (Validator validator : validators) {
            if (validator.getState() == ValidatorState.NOT_VALIDATED
                    || validator.getState() == ValidatorState.INVALID) {
                validator.validate();
            }
        }
    }

    /**
     * Perform submit action as soon as the form is {@link ValidatorState#VALID}.
     */
    public void submitWhenValid() {
        switch (getState()) {
            case NOT_VALIDATED:
            case INVALID:
                submitWhenValid = true;
                validate();
                break;
            case VALIDATING:
                submitWhenValid = true;
                break;
            case VALID:
                if (submitAction != null) {
                    submitAction.performAction();
                }
                break;
        }
    }

    public void dispose() {
        if (livDisposable != null && !livDisposable.isDisposed()) {
            livDisposable.dispose();
        }

        for (Validator validator : validators) {
            validator.dispose();
        }
    }

    public ValidatorState getState() {
        return state;
    }

    public interface Callback {
        void onStateChange(ValidatorState state);
    }

    public interface Action {
        void performAction();
    }


    public static final class Builder {
        private List<Validator> validators;
        private Callback callback;
        private Action submitAction;

        public Builder() {
            validators = new ArrayList<>();
        }

        public Builder add(Validator validator) {
            validators.add(validator);
            return this;
        }

        public Builder add(TextInputLayout input, ValidationTime time, MessageType messageType, Rule... rules) {
            add(new TextInputLayoutValidator(input, time, messageType, Arrays.asList(rules)));
            return this;
        }

        public Builder add(TextInputLayout input, ValidationTime time, Rule... rules) {
            add(new TextInputLayoutValidator(input, time, Arrays.asList(rules)));
            return this;
        }

        public Builder add(TextInputLayout input, MessageType messageType, Rule... rules) {
            add(new TextInputLayoutValidator(input, messageType, Arrays.asList(rules)));
            return this;
        }

        public Builder add(TextInputLayout input, Rule... rules) {
            add(new TextInputLayoutValidator(input, Arrays.asList(rules)));
            return this;
        }

        public Builder callback(Callback val) {
            callback = val;
            return this;
        }

        public Builder submitAction(Action val) {
            submitAction = val;
            return this;
        }

        public Liv build() {
            return new Liv(this);
        }
    }
}

File: liv/src/main/java/com/benitobertoli/liv/Liv.kt
package com.benitobertoli.liv

import com.benitobertoli.liv.scheduler.SchedulerProvider
import com.benitobertoli.liv.scheduler.SchedulerProviderImpl
import com.benitobertoli.liv.validator.Validator
import com.benitobertoli.liv.validator.ValidatorState
import com.benitobertoli.liv.validator.ValidatorState.INVALID
import com.benitobertoli.liv.validator.ValidatorState.NOT_VALIDATED
import com.benitobertoli.liv.validator.ValidatorState.VALID
import com.benitobertoli.liv.validator.ValidatorState.VALIDATING
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable


class Liv(
        private val schedulerProvider: SchedulerProvider = SchedulerProviderImpl,
        validators: List<Validator> = listOf(),
        private var submitAction: (() -> Unit)? = null,
        private var stateChanges: ((ValidatorState) -> Unit)? = null
) {
    private val validators: MutableList<Validator> = validators.toMutableList()
    private var validatorState = NOT_VALIDATED
    private var submitWhenValid = false
    private var livDisposable = CompositeDisposable()

    // Java compatibility
    constructor(builder: Builder) : this(builder.schedulerProvider, builder.validators, builder.submitAction, builder.stateChanges)

    fun add(validator: Validator) {
        validators.add(validator)
    }

    fun callback(stateChanges: ((ValidatorState) -> Unit)?) {
        this.stateChanges = stateChanges
    }

    fun submitAction(submitAction: (() -> Unit)?) {
        this.submitAction = submitAction
    }

    fun start() {
        livDisposable.add(
                Observable.combineLatest(validators.map { it.state }) { states -> resolveState(states) }
                        .subscribeOn(schedulerProvider.foregroundScheduler)
                        .observeOn(schedulerProvider.foregroundScheduler)
                        .subscribe({ state ->
                            validatorState = state
                            stateChanges?.invoke(validatorState)
                            submitAction?.let { invokeActionIfValid(it) }
                        }, { it.printStackTrace() }))
    }

    private fun resolveState(states: Array<out Any>): ValidatorState {
        var hasNotValidated = false
        var hasInvalid = false

        states.forEach { state ->
            when (state) {
                NOT_VALIDATED -> hasNotValidated = true
                VALIDATING -> return VALIDATING
                INVALID -> hasInvalid = true
                else -> {
                }
            }
        }

        if (hasNotValidated) {
            return NOT_VALIDATED
        } else if (hasInvalid) {
            return INVALID
        }

        return VALID
    }

    private fun invokeActionIfValid(action: () -> Unit) {
        if (submitWhenValid) {
            when (validatorState) {
                VALID -> {
                    submitWhenValid = false
                    action()
                }
                INVALID -> submitWhenValid = false
                else -> {
                }
            }
        }
    }

    private fun validate() {
        validators
                .filter { it.value == NOT_VALIDATED || it.value == INVALID }
                .forEach { it.validate() }
    }

    fun submitWhenValid() {
        when (validatorState) {
            NOT_VALIDATED, INVALID -> {
                submitWhenValid = true
                validate()
            }
            VALIDATING -> submitWhenValid = true
            VALID -> submitAction?.invoke()
        }
    }

    fun dispose() {
        livDisposable.dispose()

        for (validator in validators) {
            validator.dispose()
        }
    }

    // Java compatibility
    class Builder(val schedulerProvider: SchedulerProvider = SchedulerProviderImpl) {
        val validators = mutableListOf<Validator>()
        var submitAction: (() -> Unit)? = null
        var stateChanges: ((ValidatorState) -> Unit)? = null

        fun add(validator: Validator): Builder {
            validators.add(validator)
            return this
        }

        fun submitAction(action: (() -> Unit)?): Builder {
            submitAction = action
            return this
        }

        fun stateChanges(stateChanges: ((ValidatorState) -> Unit)?): Builder {
            this.stateChanges = stateChanges
            return this
        }

        fun build(): Liv {
            return Liv(this)
        }
    }
}
